### PR TITLE
Support for SAML (Touchstone) login

### PR DESCRIPTION
**Pull request by mbertrand**
#### What are the relevant tickets?
- Partially addresses #671

#### What's this PR do?
- Add support for SAML login/registration (used by Touchstone)

#### How should this be manually tested?
- `docker-compose build` (new requirements)
- Follow the steps in the [new section of the README](https://github.com/mitodl/open-discussions/tree/mb/saml#7-testing-integration-with-saml-via-ssocircle)
- Depending on the email you registered with for SSOCircle, a new user with profile should be created or you will be logged in as an existing user with the matching email address.

### CODE REVIEW

File: authentication/pipeline/user.py [hunk 1]
```diff
@@ -108,6 +109,44 @@ def require_password_and_profile(
...
 
 
+@partial
+def require_profile_update_user_via_saml(
+        strategy, backend, user=None, is_new=False, *args, **kwargs
+):  # pylint: disable=unused-argument
+    """
+    Sets a new user's password and profile, and updates the user first and last names.
+    Touchstone only returns DisplayName, which is the full name, and that is what the
+    user first_name is initially set to.
```
Comment (noisecapella) [inline authentication/pipeline/user.py]: What happens if the touchstone user's full name is greater than 30 characters (the limit on the field)? We should also check this below and add long names to our tests

File: Dockerfile [hunk 2]
```diff
@@ -14,6 +14,9 @@ RUN apt-get install -y $(grep -vE "^\s*#" apt.txt  | tr "\n" " ")
 # Add repo needed for postgres 9.6 and install it
 RUN apt-get update && apt-get install libpq-dev postgresql-client-9.6 -y
 
+# Add repo needed for python3-saml
```
Comment (noisecapella) [inline Dockerfile]: This shouldn't be necessary since you already added it to apt.txt

**Comment by noisecapella** [comment]
Just went through manual testing, everything looks great

File: authentication/pipeline/user.py (hunk 1 above)
Comment (mbertrand) [reply authentication/pipeline/user.py]: I modified the name assignment code to handle this and truncate first and last names if they are more than 30 characters; tests included.

**Comment by mbertrand** [comment]
@shaidar @pdpinch before the Touchstone login actually works, there's x509 certificates that have to be created, new environment variables that have to be set, and URL's that have to be registered with Touchstone.  Should I wait on merging this until those are in place?  Nothing should break if it's merged.

**Comment by mbertrand** [comment]
@pdpinch I'm going to go ahead and merge this, it shouldn't affect existing functionality and can be tested with Touchstone whenever we get the x509 certificates and populate the new env. variables.

File: authentication/pipeline/user.py (hunk 1 above)
Comment (pdpinch) [reply authentication/pipeline/user.py]: I'm not sure 30 characters is a reasonable limit for a full name...

Comment (mbertrand) [reply authentication/pipeline/user.py]: That's 30 characters per first name, 30 per last name: these are the defaults for the `django.contrib.auth.models.User` model (though last_name goes up to 150 in Django 2.0)

**Pull request merged**